### PR TITLE
feat(byon): Import container image pipeline

### LINKED ISSUE #22: [BYON] Create a pipeline that would serve as a byon image import pipeline
## Problem statement
Part of: https://github.com/open-services-group/byon/issues/1
Requires: https://github.com/thoth-station/helm-charts/issues/21

## High-level Goals
Import image as imagestream

## Proposal description
A pipeline that:
1. Accepts an image url, name and description as an input parameter
2. Creates an image stream populating annotations and labels:
   ```yaml
   annotations:
     opendatahub.io/notebook-image-desc: <INPUT_DESCRIPTION>
     opendatahub.io/notebook-image-name: <INPUT_NAME>
     opendatahub.io/notebook-image-phase: Validating
     opendatahub.io/notebook-image-url: <INPUT_URL>
     opendatahub.io/notebook-image-visible: "false"
   labels:
     opendatahub.io/notebook-image: 'true'
   ```
3. At the same time it initiates validation task
4. After both tasks are completed aggregate results from validation and update the image stream with new findings:
   ```yaml
   annotations:
     opendatahub.io/notebook-image-phase: Succeeded/Failed
     opendatahub.io/notebook-image-visible: 'true'  # If succeeded
   ```
   and populate the `.spec.tags[0]` with:
   ```yaml
   name: latest
      annotations:
        opendatahub.io/notebook-python-dependencies: <DISCOVERED_PACKAGES>
        opendatahub.io/notebook-software: <PYTHON_VERSION>
        openshift.io/imported-from: <IMPORT_URL>
      from:
        kind: DockerImage
        name: <IMPORT_URL>
   ```

## Acceptance Criteria

- [ ] New pipeline resource named `byon-import-jupyterhub-image`

### LINKED ISSUE #25: [BYON] Emit multiple recommendations per validation
Instead of finding a single violation and exiting the validation, continue if possible and raise as many recommendations possible. Gather Recommendations in [NDJSON](http://ndjson.org/) format, so it can be easily serialized later in an imagestream update step. Message format:

```
{"severity": "error", "message": "Python not found"}
{"severity": "warning", "message": "Package 'non-critical-package' not installed"}
```

Action item: decide if validation should fail on any error (including recommendations) or we should let the validation pass if only certain severity is present.
Note: we may need to use workspaces to pass extra data, since Tekton has results limit size at `4096 bytes` https://tekton.dev/docs/pipelines/tasks/#emitting-results

### PR DESCRIPTION
**Pull request by tumido**
Resolves: https://github.com/thoth-station/helm-charts/issues/22 
Resolves: https://github.com/thoth-station/helm-charts/issues/25 

Examples:

#### Image passes validation

https://user-images.githubusercontent.com/7453394/153030337-999dc45f-6620-44dd-bd8f-c4bab66ed77f.mp4

#### Image fails validation

https://user-images.githubusercontent.com/7453394/153030329-52325b11-e42d-4040-b511-7e22fca523bd.mp4

### CODE REVIEW

File: charts/meteor-pipelines/templates/byon-import-jupyterhub-image.yaml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            echo "  Image name:        $(params.name)"
+            echo "  Image description: $(params.desc)"
+            echo "  Image URL:         $(params.url)"
+            echo
+            echo "  Phase:             Validating"
+            echo "  Message:           ''"
+            echo "  Visibility:        false"
+            echo
+
+            cat <<EOM | oc apply -f -
```
Comment (codificat) [inline charts/meteor-pipelines/templates/byon-import-jupyterhub-image.yaml]: [UUOC](https://en.wikipedia.org/wiki/Cat_(Unix)#Useless_use_of_cat)? I believe this is better:

```suggestion
            oc apply -f - <<EOM
```

**Review by codificat — COMMENTED** [review]
This looks great, thanks.
Only a minor suggestion for now

File: charts/meteor-pipelines/templates/byon-import-jupyterhub-image.yaml (hunk 1 above)
Comment (tumido) [reply charts/meteor-pipelines/templates/byon-import-jupyterhub-image.yaml]: That's fair, yup :slightly_smiling_face:

**Comment by tumido** [comment]
New iteration with #21 included. Videos above updated!

We've had to switch over to use a `workspace` to pass data between tasks. Using `output`s can make workflow to skip a task in `finally` in case this task depends on an output of the failed task... We don't want this to ever happen, the `finally` tasks must be always executed.

**Comment by tumido** [comment]
@codificat can you help me please? Do you have any idea why is the linter failing here?
image

**Comment by codificat** [comment]
> @codificat can you help me please? Do you have any idea why is the linter failing here?

Unfortunately the tests are still a bit flaky :disappointed:   e.g. sometimes the tekton webhook fails to register on the microshift cluster; I have not found what's going on in these situations yet.

I just re-ran the jobs and it's green now. A `/retest` should help if this happens again. Besides that,
/lgtm

Feel free to /assign someone for final approval (it can be myself, just wondering if you want another review here)

**Comment by codificat** [comment]
I tested this locally using https://github.com/open-services-group/byon/pull/21 and I hit a problem obtaining the details of the python env and deps:

```
pep@uio byon (quick-dev-env) $ make run
tkn pipeline start byon-import-jupyterhub-image \
	--showlog \
	-w name=data,volumeClaimTemplateFile=manifests/overlays/dev/pvc.yaml \
	-p url=quay.io/tcoufal/false \
	-p name="This will FAIL" \
	-p desc="This image is expected to FAIL validation"
PipelineRun started: byon-import-jupyterhub-image-run-x7l9j
Waiting for logs to be available...
[create-imagestream : oc] Creating image stream:
[create-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-x7l9j
[create-imagestream : oc]   Namespace:  tumido
[create-imagestream : oc] 
[create-imagestream : oc]   Image name:        This will FAIL
[create-imagestream : oc]   Image description: This image is expected to FAIL validation
[create-imagestream : oc]   Image URL:         quay.io/tcoufal/false
[create-imagestream : oc] 
[create-imagestream : oc]   Phase:             Validating
[create-imagestream : oc]   Message:           ''
[create-imagestream : oc]   Visibility:        false
[create-imagestream : oc] 
[create-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-x7l9j created

[setup : setup] Creating output files... DONE
[setup : setup] Verifying image is pull-able... FAILED

failed to get logs for task setup : container step-setup has failed  : [{"key":"StartedAt","value":"2022-02-09T12:54:57.697Z","type":3}]
[update-imagestream : oc] Updating image stream:
[update-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-x7l9j
[update-imagestream : oc]   Namespace:  tumido
[update-imagestream : oc] 
[update-imagestream : oc]   Image tag:           latest
[update-imagestream : oc]   Phase:               Failed
[update-imagestream : oc]   Message:             [{"severity":"error","message":"Unable to pull image"}]
[update-imagestream : oc]   Visibility:          false
[update-imagestream : oc]   Software:            []
[update-imagestream : oc]   Python dependencies: []
[update-imagestream : oc] 
[update-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-x7l9j patched

tkn pipeline start byon-import-jupyterhub-image \
	--showlog \
	-w name=data,volumeClaimTemplateFile=manifests/overlays/dev/pvc.yaml \
	-p url=quay.io/thoth-station/s2i-minimal-notebook:v0.0.15 \
	-p name="This will PASS" \
	-p desc="This image is expected to PASS validation"
PipelineRun started: byon-import-jupyterhub-image-run-hqx9d
Waiting for logs to be available...
[setup : setup] Creating output files... DONE
[create-imagestream : oc] Creating image stream:
[create-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-hqx9d
[create-imagestream : oc]   Namespace:  tumido
[create-imagestream : oc] 
[create-imagestream : oc]   Image name:        This will PASS
[create-imagestream : oc]   Image description: This image is expected to PASS validation
[create-imagestream : oc]   Image URL:         quay.io/thoth-station/s2i-minimal-notebook:v0.0.15
[create-imagestream : oc] 
[create-imagestream : oc]   Phase:             Validating
[create-imagestream : oc]   Message:           ''
[create-imagestream : oc]   Visibility:        false
[create-imagestream : oc] 
[setup : setup] Verifying image is pull-able... DONE

[create-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-hqx9d created

[validate : jupyter-book-build] Validating image:
[validate : jupyter-book-build]     Python:     3.6.8
[validate : jupyter-book-build] /tekton/scripts/script-0-62prs: line 8: /workspace/data/notebook-software.json: Permission denied
[validate : jupyter-book-build] /tekton/scripts/script-0-62prs: line 17: /workspace/data/notebook-python-dependencies.json: Permission denied
[validate : jupyter-book-build]     Packages:   jupyter-kernel-gateway==2.4.0 jupyter-nbrequirements==0.7.3 jupyterhub==1.3.0 jupyterlab==3.0.14 jupyterlab-git==0.30.0b2 mod-wsgi==4.6.8 pipenv==2020.11.15 supervisor==4.1.0

[update-imagestream : oc] Updating image stream:
[update-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-hqx9d
[update-imagestream : oc]   Namespace:  tumido
[update-imagestream : oc] 
[update-imagestream : oc]   Image tag:           v0.0.15
[update-imagestream : oc]   Phase:               Succeeded
[update-imagestream : oc]   Message:             
[update-imagestream : oc]   Visibility:          true
[update-imagestream : oc]   Software:            []
[update-imagestream : oc]   Python dependencies: []
[update-imagestream : oc] 
[update-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-hqx9d patched
```

**Comment by tumido** [comment]
That is confusing.. there's nothing that should effect permissions in there. And based on the logs the files are available, they were created in th `setup` step

https://github.com/thoth-station/helm-charts/pull/23/files#diff-502c30bc43b0de34014a4c791a3b5480a7666fb05f7346476a447ed04eee1e20R82-R83

You're getting permission errors on these 2 lines:
https://github.com/thoth-station/helm-charts/pull/23/files#diff-c4b699746e2ca34e4a1d2cd06915c1cda93670cb2200bc77f55534db191abf5bR23
https://github.com/thoth-station/helm-charts/pull/23/files#diff-c4b699746e2ca34e4a1d2cd06915c1cda93670cb2200bc77f55534db191abf5bR32

Which is very weird since both steps are using the same settings when accessing workspaces. :shrug:  Sigh.. Back to the drawing board :smile:

**Comment by tumido** [comment]
Just to make things even weirder I'm unable to reproduce this. What is the storage backend on your dev cluster @codificat ?

**Comment by codificat** [comment]
> Just to make things even weirder I'm unable to reproduce this. What is the storage backend on your dev cluster @codificat ?

NFS.

I did not have time yet to look at the issue itself, will take a look at the PV now

**Comment by codificat** [comment]
The difference I see is that the setup step runs from the `skopeo` container image, which runs as root, whereas the validation runs from the to-validate image, which in this case runs with uid 1001.

In my PV, the files created during setup are owned by root and only writtable by root.

**Comment by tumido** [comment]
Interesting.. It is very weird that the other workflow executed just fine for you. Because in both cases it needs to be able to access the workspace.

In the first run the `{"severity":"error","message":"Unable to pull image"}` message was successfully read from `messages.ndjson` but in the second run it failed to read from `notebook-software.json` and `notebook-python-dependencies.json` in the same folder. The mount is the same, the process (and user) creating all the files is the same,.. I don't see why it can fail.

This is what I'm seeing in a successful run:

```
$ oc get pods -l tekton.dev/pipelineRun=byon-import-jupyterhub-image-run-ln2qj
NAME                                                              READY   STATUS      RESTARTS   AGE
byon-import-jupyterhub-image-run-ln2qj-create-imagestream-jzqg5   0/1     Completed   0          19h
byon-import-jupyterhub-image-run-ln2qj-setup-96spj-pod-nvr6m      0/1     Completed   0          19h
byon-import-jupyterhub-image-run-ln2qj-update-imagestream-t79zp   0/1     Completed   0          19h
byon-import-jupyterhub-image-run-ln2qj-validate-8k7th-pod-vvhfr   0/1     Completed   0          19h

$ oc debug byon-import-jupyterhub-image-run-ln2qj-setup-96spj-pod-nvr6m                                                            
Starting pod/byon-import-jupyterhub-image-run-ln2qj-setup-96spj-pod-nvr6m-debug, command was: /tekton/tools/entrypoint -wait_file /tekton/downward/ready -wait_file_content -post_file /tekton/tools/0 -termination_path /tekton/termination -step_metadata_dir /tekton/steps/step-setup -step_metadata_dir_link /tekton/steps/0 -docker-cfg=pipeline-dockercfg-c4bbk -entrypoint /tekton/scripts/script-0-hfqbx --
iPod IP: 10.129.2.221
If you don't see a command prompt, try pressing enter.
sh-4.4# id
uid=0(root) gid=0(root) groups=0(root),1000660000
sh-4.4# ls -l /workspace/data/
total 28
drwxrws---. 2 root 1000660000 16384 Feb  9 13:53 lost+found
-rw-rw-r--. 1 root 1000660000     0 Feb  9 13:53 messages.ndjson
-rw-rw-r--. 1 root 1000660000   383 Feb  9 13:53 notebook-python-dependencies.json
-rw-rw-r--. 1 root 1000660000    37 Feb  9 13:53 notebook-software.json
-rw-rw-r--. 1 root 1000660000   976 Feb  9 13:54 patch-file.yaml
sh-4.4# exit

Removing debug pod ...

$ oc debug byon-import-jupyterhub-image-run-ln2qj-validate-8k7th-pod-vvhfr
Starting pod/byon-import-jupyterhub-image-run-ln2qj-validate-8k7th-pod-vvhfr-debug, command was: /tekton/tools/entrypoint -wait_file /tekton/downward/ready -wait_file_content -post_file /tekton/tools/0 -termination_path /tekton/termination -step_metadata_dir /tekton/steps/step-jupyter-book-build -step_metadata_dir_link /tekton/steps/0 -docker-cfg=pipeline-dockercfg-c4bbk -entrypoint /tekton/scripts/script-0-4zq5b --
Pod IP: 10.129.2.222
If you don't see a command prompt, try pressing enter.
(app-root) sh-4.4$ id
uid=1001(default) gid=0(root) groups=0(root),1000660000
(app-root) sh-4.4$ ls -l /workspace/data/
total 28
drwxrws---. 2 root 1000660000 16384 Feb  9 13:53 lost+found
-rw-rw-r--. 1 root 1000660000     0 Feb  9 13:53 messages.ndjson
-rw-rw-r--. 1 root 1000660000   383 Feb  9 13:53 notebook-python-dependencies.json
-rw-rw-r--. 1 root 1000660000    37 Feb  9 13:53 notebook-software.json
-rw-rw-r--. 1 root 1000660000   976 Feb  9 13:54 patch-file.yaml
(app-root) sh-4.4$ exit

Removing debug pod ...

$ oc debug byon-import-jupyterhub-image-run-ln2qj-update-imagestream-t79zp
Starting pod/byon-import-jupyterhub-image-run-ln2qj-update-imagestream-t79zp-debug, command was: /tekton/tools/entrypoint -wait_file /tekton/downward/ready -wait_file_content -post_file /tekton/tools/0 -termination_path /tekton/termination -step_metadata_dir /tekton/steps/step-oc -step_metadata_dir_link /tekton/steps/0 -docker-cfg=pipeline-dockercfg-c4bbk -entrypoint /tekton/scripts/script-0-6gr7x --
Pod IP: 10.131.0.163
If you don't see a command prompt, try pressing enter.
sh-4.4# id
uid=0(root) gid=0(root) groups=0(root),1000660000
sh-4.4# ls -l /workspace/manifest-dir/
total 28
drwxrws---. 2 root 1000660000 16384 Feb  9 13:53 lost+found
-rw-rw-r--. 1 root 1000660000     0 Feb  9 13:53 messages.ndjson
-rw-rw-r--. 1 root 1000660000   383 Feb  9 13:53 notebook-python-dependencies.json
-rw-rw-r--. 1 root 1000660000    37 Feb  9 13:53 notebook-software.json
-rw-rw-r--. 1 root 1000660000   976 Feb  9 13:54 patch-file.yaml
sh-4.4# exit

Removing debug pod ...
```

See the `uid` change - as you've said, it differs, however in my storage backend (`cider`) it causes no troubles :confused: Testing `nfs` now

**Comment by codificat** [comment]
Thanks for the details.

The difference I see in my PV is the permissions on the files: in your case it still works as both containers run with gid=0 and the files have group write perm.

In my PV:

```
$ oc debug byon-import-jupyterhub-image-run-hqx9d-validate-jhflv-pod-2k8b9
Starting pod/byon-import-jupyterhub-image-run-hqx9d-validate-jhflv-pod-2k8b9-debug, command was: /tekton/tools/entrypoint -wait_file /tekton/downward/ready -wait_file_content -post_file /tekton/tools/0 -termination_path /tekton/termination -step_metadata_dir /tekton/steps/step-jupyter-book-build -step_metadata_dir_link /tekton/steps/0 -docker-cfg=pipeline-dockercfg-ctx59 -entrypoint /tekton/scripts/script-0-62prs --
Pod IP: 10.131.1.213
If you don't see a command prompt, try pressing enter.
(app-root) sh-4.4$ id
uid=1001(default) gid=0(root) groups=0(root),1000680000
(app-root) sh-4.4$ ls -l /workspace/data/
total 12
-rw-r--r--. 1 root root   0 Feb  9 12:55 messages.ndjson
-rw-r--r--. 1 root root   2 Feb  9 12:55 notebook-python-dependencies.json
-rw-r--r--. 1 root root   2 Feb  9 12:55 notebook-software.json
-rw-r--r--. 1 root root 560 Feb  9 12:56 patch-file.yaml
```

What sets the umask in these pods?

**Comment by codificat** [comment]
> The difference I see in my PV is the permissions on the files: in your case it still works as both containers run with gid=0 and the files have group write perm.

Well, not exactly: I see that your files are not owned by gid 0 (they are in mine) - but both pods share gid 1000660000 and the files have 0664 perms

**Comment by tumido** [comment]
I think https://github.com/thoth-station/helm-charts/pull/23/commits/95be35e3fc1fd00a2ba4db30b56c33ec6d420e82 should fix our permission issue. Files created in setup step in workspace should be "read-write for all".

**Comment by codificat** [comment]
> I think [95be35e](https://github.com/thoth-station/helm-charts/commit/95be35e3fc1fd00a2ba4db30b56c33ec6d420e82) should fix our permission issue. Files created in setup step in workspace should be "read-write for all".

This certainly did the trick, thanks.

/approve

<details>
<summary>For the record, last test here:</summary>

```
pep@uio byon (quick-dev-env) $ make run
tkn pipeline start byon-import-jupyterhub-image \
	--showlog \
	-w name=data,volumeClaimTemplateFile=manifests/overlays/dev/pvc.yaml \
	-p url=quay.io/tcoufal/false \
	-p name="This will FAIL" \
	-p desc="This image is expected to FAIL validation"
PipelineRun started: byon-import-jupyterhub-image-run-g8wvg
Waiting for logs to be available...
[create-imagestream : oc] Creating image stream:
[create-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-g8wvg
[create-imagestream : oc]   Namespace:  tumido
[create-imagestream : oc] 
[create-imagestream : oc]   Image name:        This will FAIL
[create-imagestream : oc]   Image description: This image is expected to FAIL validation
[create-imagestream : oc]   Image URL:         quay.io/tcoufal/false
[create-imagestream : oc] 
[create-imagestream : oc]   Phase:             Validating
[create-imagestream : oc]   Message:           ''
[create-imagestream : oc]   Visibility:        false
[create-imagestream : oc] 
[setup : setup] Creating output files... DONE
[setup : setup] Verifying image is pull-able... FAILED
[create-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-g8wvg created

failed to get logs for task setup : container step-setup has failed  : [{"key":"StartedAt","value":"2022-02-11T17:59:35.304Z","type":3}]

[update-imagestream : oc] Updating image stream:
[update-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-g8wvg
[update-imagestream : oc]   Namespace:  tumido
[update-imagestream : oc] 
[update-imagestream : oc]   Image tag:           latest
[update-imagestream : oc]   Phase:               Failed
[update-imagestream : oc]   Message:             [{"severity":"error","message":"Unable to pull image"}]
[update-imagestream : oc]   Visibility:          false
[update-imagestream : oc]   Software:            []
[update-imagestream : oc]   Python dependencies: []
[update-imagestream : oc] 
[update-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-g8wvg patched

tkn pipeline start byon-import-jupyterhub-image \
	--showlog \
	-w name=data,volumeClaimTemplateFile=manifests/overlays/dev/pvc.yaml \
	-p url=quay.io/thoth-station/s2i-minimal-notebook:v0.0.15 \
	-p name="This will PASS" \
	-p desc="This image is expected to PASS validation"
PipelineRun started: byon-import-jupyterhub-image-run-9rwpd
Waiting for logs to be available...
[create-imagestream : oc] Creating image stream:
[create-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-9rwpd
[create-imagestream : oc]   Namespace:  tumido
[create-imagestream : oc] 
[create-imagestream : oc]   Image name:        This will PASS
[create-imagestream : oc]   Image description: This image is expected to PASS validation
[create-imagestream : oc]   Image URL:         quay.io/thoth-station/s2i-minimal-notebook:v0.0.15
[create-imagestream : oc] 
[create-imagestream : oc]   Phase:             Validating
[create-imagestream : oc]   Message:           ''
[create-imagestream : oc]   Visibility:        false
[create-imagestream : oc] 
[setup : setup] Creating output files... DONE
[setup : setup] Verifying image is pull-able... DONE
[create-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-9rwpd created


[validate : jupyter-book-build] Validating image:
[validate : jupyter-book-build]     Python:     3.6.8
[validate : jupyter-book-build]     Packages:   jupyter-kernel-gateway==2.4.0 jupyter-nbrequirements==0.7.3 jupyterhub==1.3.0 jupyterlab==3.0.14 jupyterlab-git==0.30.0b2 mod-wsgi==4.6.8 pipenv==2020.11.15 supervisor==4.1.0

[update-imagestream : oc] Updating image stream:
[update-imagestream : oc]   Name:       byon-import-jupyterhub-image-run-9rwpd
[update-imagestream : oc]   Namespace:  tumido
[update-imagestream : oc] 
[update-imagestream : oc]   Image tag:           v0.0.15
[update-imagestream : oc]   Phase:               Succeeded
[update-imagestream : oc]   Message:             
[update-imagestream : oc]   Visibility:          true
[update-imagestream : oc]   Software:            [{"name":"Python","version":"3.6.8"}]
[update-imagestream : oc]   Python dependencies: [{"name": "jupyter-kernel-gateway", "version": "2.4.0"}, {"name": "jupyter-nbrequirements", "version": "0.7.3"}, {"name": "jupyterhub", "version": "1.3.0"}, {"name": "jupyterlab", "version": "3.0.14"}, {"name": "jupyterlab-git", "version": "0.30.0b2"}, {"name": "mod-wsgi", "version": "4.6.8"}, {"name": "pipenv", "version": "2020.11.15"}, {"name": "supervisor", "version": "4.1.0"}]
[update-imagestream : oc] 
[update-imagestream : oc] imagestream.image.openshift.io/byon-import-jupyterhub-image-run-9rwpd patched
```

And the files in the PV:

```
# ls -l /srv/ocpstorage/tumido-pvc-*
/srv/ocpstorage/tumido-pvc-3bd0e554bb-pvc-5c739338-750f-42b6-931d-86ec1994003a:
total 12
-rw-rw-rw-. 1 root root   0 Feb 11 12:59 messages.ndjson
-rw-rw-rw-. 1 root root 383 Feb 11 12:59 notebook-python-dependencies.json
-rw-rw-rw-. 1 root root  37 Feb 11 12:59 notebook-software.json
-rw-r--r--. 1 root root 976 Feb 11 12:59 patch-file.yaml

/srv/ocpstorage/tumido-pvc-58c0b89854-pvc-49236253-0da3-4235-92b7-40370bbc39c8:
total 16
-rw-rw-rw-. 1 root root  54 Feb 11 12:59 messages.ndjson
-rw-rw-rw-. 1 root root   2 Feb 11 12:59 notebook-python-dependencies.json
-rw-rw-rw-. 1 root root   2 Feb 11 12:59 notebook-software.json
-rw-r--r--. 1 root root 554 Feb 11 12:59 patch-file.yaml
```
</details>

**Pull request merged**